cmd/server.go

```diff
@@ -0,0 +1,123 @@
+package cmd
+
+import (
+	"flag"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/skuid/keyman/oidcauth"
+	"github.com/skuid/keyman/server"
+	"github.com/skuid/keyman/sign"
+	"github.com/skuid/spec"
+	"github.com/skuid/spec/lifecycle"
+	_ "github.com/skuid/spec/metrics" // import spec metrics
+	"github.com/skuid/spec/middlewares"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var level = zapcore.InfoLevel
+
+var serverCmd = &cobra.Command{
+	Use:   "server",
+	Short: "Run a SSH key signing server",
+	// Set up logging
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		l, err := spec.NewStandardLevelLogger(level)
+		if err != nil {
+			return fmt.Errorf("Error initializing logger: %q", err)
+		}
+		zap.ReplaceGlobals(l)
+		return nil
+	},
+	Run: serverFunc,
+}
+
+func serverFunc(cmd *cobra.Command, args []string) {
+
+	ca, err := sign.ReadPrivKey(viper.GetString("key"))
+	if err != nil {
+		zap.L().Fatal("Error reading key", zap.Error(err))
+	}
+
+	authority := &server.Authority{
+		CA:             ca,
+		CaComment:      viper.GetString("ca-name"),
+		Duration:       viper.GetDuration("validity"),
+		IdentityHeader: oidcauth.Identity,
+	}
+	hostPort := fmt.Sprintf(":%d", viper.GetInt("port"))
+
+	// Metrics and lifecycle server
+	go func() {
+		internalMux := http.NewServeMux()
+		internalMux.Handle("/metrics", promhttp.Handler())
+		internalMux.HandleFunc("/live", lifecycle.LivenessHandler)
+		internalMux.HandleFunc("/ready", lifecycle.ReadinessHandler)
+		metricsHostPort := fmt.Sprintf(":%d", viper.GetInt("metrics-port"))
+
+		zap.L().Info("Metrics server is starting", zap.String("listen", metricsHostPort))
+		httpServer := &http.Server{Addr: metricsHostPort, Handler: internalMux}
+		lifecycle.ShutdownOnTerm(httpServer)
+		if err := httpServer.ListenAndServe(); err != http.ErrServerClosed {
+			zap.L().Fatal("Error listening", zap.Error(err))
+		}
+		zap.L().Info("Metrics server gracefully stopped")
+	}()
```

Comment:
This is a good pattern that we could probably drop into the lifecycle package of the spec at some point. We serve these routes off of the same server as the app routes in warden, but it makes sense to just run another server independently.

Reply:
Yea that is a good idea. I'll update skuid/spec#12 with it